python/ray/serve/tests/unit/test_deployment_rank_manager.py

```diff
@@ -0,0 +1,442 @@
+        rank_manager = DeploymentRankManager(_fail_on_error=True)
+        replica1 = MockDeploymentReplica("replica_1")
+
+        # Set up stale rank (replica not in active list)
+        rank_manager.assign_rank("replica_1")
+        rank_manager.assign_rank("stale_replica")
+
+        with pytest.raises(
+            RuntimeError, match="Controller rank system is in an invalid state"
+        ):
+            rank_manager.check_rank_consistency_and_reassign_minimally([replica1])
+
+    def test_check_rank_consistency_stale_ranks_logs_when_flag_disabled(self):
+        """Test consistency check only logs when there are stale ranks and flag is disabled."""
+        rank_manager = DeploymentRankManager(_fail_on_error=False)
+        replica1 = MockDeploymentReplica("replica_1")
+
+        # Set up stale rank (replica not in active list)
+        rank_manager.assign_rank("replica_1")
+        rank_manager.assign_rank("stale_replica")
+
+        # When flag is disabled, it logs error but continues with reassignment
+        # Since only replica_1 is active and has rank 0, no reassignment needed
+        result = rank_manager.check_rank_consistency_and_reassign_minimally([replica1])
+        assert result == []
+
+    def test_check_rank_consistency_duplicate_ranks_fails_when_flag_enabled(self):
+        """Test consistency check fails when there are duplicate ranks and flag is enabled."""
+        rank_manager = DeploymentRankManager(_fail_on_error=True)
+        replica1 = MockDeploymentReplica("replica_1")
+        replica2 = MockDeploymentReplica("replica_2")
+
+        # Manually create duplicate ranks (this should never happen in normal operation)
+        rank_manager._replica_ranks = {"replica_1": 0, "replica_2": 0}  # Duplicate!
+
+        with pytest.raises(
+            RuntimeError, match="Controller rank system is in an invalid state"
+        ):
+            rank_manager.check_rank_consistency_and_reassign_minimally(
+                [replica1, replica2]
+            )
+
+    def test_check_rank_consistency_duplicate_ranks_logs_when_flag_disabled(self):
+        """Test consistency check only logs when there are duplicate ranks and flag is disabled."""
+        rank_manager = DeploymentRankManager(_fail_on_error=False)
+        replica1 = MockDeploymentReplica("replica_1")
+        replica2 = MockDeploymentReplica("replica_2")
+
+        # Manually create duplicate ranks (this should never happen in normal operation)
+        rank_manager._replica_ranks = {"replica_1": 0, "replica_2": 0}  # Duplicate!
+        rank_manager._next_rank = 1
+
+        # When flag is disabled, it logs error but still performs reassignment to fix the issue
+        result = rank_manager.check_rank_consistency_and_reassign_minimally(
+            [replica1, replica2]
+        )
+        assert result == [replica2] or result == [replica1]
+
+
+class TestDeploymentRankManagerIntegration:
```

Comment:
let's move this to `test_deployment_state`? then it can be a more "e2e" unit test, we can test the same logic but through interacting with a mock deployment state manager